File: code/2784.check-if-array-is-good.cpp
/*
 * @lc app=leetcode id=2784 lang=cpp
 *
 * [2784] Check if Array is Good
 */
#include <vector>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
public:
    bool isGood(vector<int> &nums)
    {
        if (nums.size() <= 1)
            return false;
        int p = 1;
        int n = nums.size() - 1;
        int tmp;
        int val;
        while (p < nums.size())
        {
            val = nums[p];
            while (val != p)
            {
                if (!(val >= 0 && val <= n))
                    return false;
                tmp = nums[val];
                // place target
                if (val == n)
                {
                    if (nums[0] == n && nums[n] == n)
                        return false;
                    if (nums[n] != n)
                    {
                        tmp = nums[n];
                        nums[n] = n;
                        val = tmp;
                    }
                    else
                    {
                        tmp = nums[0];
                        nums[0] = n;
                        val = tmp;
                    }
                }
                else
                {
                    // already equal
                    if (tmp == val)
                        return false;
                    nums[val] = val;
                    val = tmp;
                }
            }
            p++;
        }
        return nums[0] == n;
    }
};
// @lc code=end

File: code/2784.check-if-array-is-good.go
/*
 * @lc app=leetcode id=2784 lang=golang
 *
 * [2784] Check if Array is Good
 */

// @lc code=start
func isGood(nums []int) bool {
	p := 1
	length := len(nums)
	val := 0
	n := length - 1
	tmp := 0
	for p < length {
		val = nums[p]
		for val != p {
			if !(val >= 0 && val < length) {
				return false
			}
			tmp = nums[val]
			if tmp == val {
				if tmp == n {
					if nums[n] != n {
						tmp = nums[n]
						nums[n] = n
					} else if nums[0] != n {
						tmp = nums[0]
						nums[0] = n
					} else {
						return false
					}
				} else {
					return false
				}
			} else {
				nums[val] = val
			}
			val = tmp
		}
		p++
	}
	return nums[0] == n
}

// @lc code=end

